Update to mui v5 from material-ui v4

illume opened the pull request:
This upgrades material-ui v4 to mui5. Additionally it provides an upgrade script for plugin authors to automate the process of updating their plugins.

- [X] Sidebar doesn't collapse properly
- [X] Styles are ported to the v5 flavor
- [x] Select, TextField, and FormControl components show the same style (not contained) as before
- [x] The logo's sidebar button (in mobile view) is not showing up
- [x] Autocomplete filter (like on Namespaces view) is broken, doesn't filter.
- [x] get snapshot testing working with generated mui class names
- [x] update snapshots
- [x] `headlamp-plugin` webpack config can work with mui v5 (add similar existing config with new mui paths)
- [x] ~~consider keeping material-ui v4 available to plugins for backwards compat. (keep exporting them for plugins, and have material ui 5 working). Seems to work so far in my testing.~~ It was decided to just support mui 5.

--- 

- [x] next round of mui5 fixes
  - [x] rebase
  - [x] add https://mui.com/material-ui/migration/migration-v4/ to commit 
  - [x] update mui packages to latest versions
  - [x] cod mod new files
  - [x] fix typescript issues
  - [x] fix test failures / typecheck errors
  - [x] Fix TopBar background color in dark mode
  - [x] fix broken plugin tests
    - [x] update plugins/index.ts to remove importing material ui.
    - [x] headlamp-plugin updates
    - [x] tests pass
  - [x] remove material-ui
  - [x] Bring in other commits from old PR
  - [x] `headlamp-plugin upgrade` should also update user plugins to mui v5. Add codemods headlamp-plugin
  - [x] do headlamp-plugin alpha release 0.8.0-alpha.0
  - [x] upgrade example plugins,` and retest them all.
  - [x] In dark theme, the links always have the underline


--- 

Whilst testing plugin upgrades, some fixes and improvements were required.

- [x] improvements to `headlamp-plugin upgrade` 
  - [x] do package upgrades more robustly.
  - [x] documentation for `headlamp-plugin upgrade`
  - [x] have an alpha tag, and a command line option to select it, so it is easier to test headlamp-plugin alpha pre-releases.
  - [x] provide migrations for a few typing issues with mui and plugins (with SVG, and package upgrades) found testing plugins.
- [x] a compatibility check. (disabled for now, because there is quite a level of compatibility already, TBD).


---

Visual UI issues remaining...

- [x] blue text for custom cluster chooser (default one is fine)
- [x] back button is on /settings is blue
- [x] documentation tab on editor is broken
- [x] switch border color is blue, on /settings theme switch
- [x] the icons are slightly larger (app bar notification icon and /settings page icon), maybe not a problem.

---

  - [x] The cancel button in the snackbars should be red-ish, not purple
  ![Snackbar with purple button screenshot](https://github.com/headlamp-k8s/headlamp/assets/1029635/819cd585-34b3-4a0d-b7b0-41ca9610e844)
  - [x] The main color is apparently now black but should be blue as before. (See all switches/buttons for example)


## How to use

Update packages and run headlamp again.

```bash
cd frontend
npm i
cd ..
make run-frontend
```

## Testing done

- [x] `npm run lint` passes
- [x] `npm run tsc` passes
- [x] `npm run test` passes
- [x] headlamp-plugin tests pass 
- [x] smoke testing
- [x] updating snapshot tests 
- [x] test old example plugins still work
- [x] test example plugins `headlamp-plugin upgrade`
- [x] test plugins repo plugins `headlamp-plugin upgrade`

illume commented:
@joaquimrocha I fixed the type checking, testing, and format checking. The Build Frontend job is now failing on the plugin tests.

illume commented:
wip

vyncent-t commented:
Good stuff!
I was browsing around and noticed there is a weird bug that causes the documentation tab when editing a resource to render a string and not the clickable elements

![image](https://github.com/headlamp-k8s/headlamp/assets/78232183/ec2580b8-1dc1-48da-a157-bd0f9552861d)

illume commented:
- Updated a few things so the `headlamp-plugin upgrade` works for all example plugins, the ai, prometheus, and app-catalog plugins.
- Added a modified compatibility check (disabled for now, more on that later).
- rebased on main

Then I did a larger side by side visual comparison between main and this branch and found these:

- [ ] blue cluster chooser
- [ ] back button is on /settings is blue
- [ ] documentation tab on editor is broken
- [ ] switch border color is blue, on /settings theme switch
- [ ] the icons are slightly larger (app bar notification icon and /settings page icon), maybe not a problem.
- [ ] in dark mode, app bar is a dark grey rather than completely black
- [ ] action icons spacing is a bit off on pod detail view /c/minikubeThisisreallyLong/pods/default/keycloak-7b478749bc-ghvdn

Now I'm working on these visual issues.

joaquimrocha reviewed (approved):
Left a question about a log. Overall looks good. I will sign my commits and also rebase the PR.

illume commented on frontend/src/plugin/index.ts:
```diff
@@ -93,39 +93,72 @@ export async function initializePlugins() {
+    }
+    return true;
+  });
+  console.log('filterSources compatible', compatible);
```
Thanks, fixed.

illume commented on frontend/src/plugin/index.ts:
```diff
@@ -93,39 +93,72 @@ export async function initializePlugins() {
+      true;
+    return enabledInSettings;
   });
+  console.log('filterSources enabledSourcesAndPackageInfos', enabledSourcesAndPackageInfos);
```
Thanks, fixed.

illume commented:
@joaquimrocha I removed the console.log's and added a couple of missing signed-off-bys.

illume commented:
I'll look at these two now...
- The cancel button in the snackbars should be red-ish, not purple
- The main color is apparently now black but should be blue as before. (See all switches/buttons for example)

illume commented:
I pushed changes for these two...
- The cancel button in the snackbars should be red-ish, not purple
- The main color is apparently now black but should be blue as before. (See all switches/buttons for example)

The back button, and the theme change button are colored black & white individually now. So they look as they did before.

joaquimrocha commented:
Thanks @illume , it's looking fine to me now! Merging now!
